contracts/contracts/strategies/BaseConvexMetaStrategy.sol

```diff
@@ -0,0 +1,218 @@
+            initState.vaultAddress,
+            _rewardTokenAddresses,
+            _assets,
+            _pTokens
+        );
+        _approveBase();
+    }
+
+    /**
+     * @dev Get the total asset value held in the platform
+     * @param _asset      Address of the asset
+     * @return balance    Total value of the asset in the platform
+     */
+    function checkBalance(address _asset)
+        public
+        view
+        virtual
+        override
+        returns (uint256 balance)
+    {
+        require(assetToPToken[_asset] != address(0), "Unsupported asset");
+        balance = 0;
+
+        // LP tokens in this contract. This should generally be nothing as we
+        // should always stake the full balance in the Gauge, but include for
+        // safety
+        uint256 contractPTokens = IERC20(pTokenAddress).balanceOf(
+            address(this)
+        );
+        ICurvePool curvePool = ICurvePool(platformAddress);
+        if (contractPTokens > 0) {
+            uint256 virtual_price = curvePool.get_virtual_price();
+            uint256 value = contractPTokens.mulTruncate(virtual_price);
+            balance += value;
+        }
+
+        /* We intentionally omit the metapoolLp tokens held by the metastrategyContract
+         * since the contract should never (except in the middle of deposit/withdrawal
+         * transaction) hold any amount of those tokens in normal operation. There
+         * could be tokens sent to it by a 3rd party and we decide to actively ignore
+         * those.
+         */
+        uint256 metapoolGaugePTokens = IRewardStaking(cvxRewardStakerAddress)
+            .balanceOf(address(this));
+
+        if (metapoolGaugePTokens > 0) {
+            uint256 value = metapoolGaugePTokens.mulTruncate(
+                metapool.get_virtual_price()
+            );
+            balance += value;
+        }
+
+        uint256 assetDecimals = Helpers.getDecimals(_asset);
+        balance = balance.scaleBy(assetDecimals, 18) / THREEPOOL_ASSET_COUNT;
+    }
+
+    function _approveBase() internal override {
+        IERC20 pToken = IERC20(pTokenAddress);
+        // 3Pool for LP token (required for removing liquidity)
+        pToken.safeApprove(platformAddress, 0);
+        pToken.safeApprove(platformAddress, type(uint256).max);
+        // Gauge for LP token
+        metapoolLPToken.safeApprove(cvxDepositorAddress, 0);
+        metapoolLPToken.safeApprove(cvxDepositorAddress, type(uint256).max);
+        // Metapool for LP token
+        pToken.safeApprove(address(metapool), 0);
+        pToken.safeApprove(address(metapool), type(uint256).max);
+        // Metapool for Metapool main token
+        metapoolMainToken.safeApprove(address(metapool), 0);
+        metapoolMainToken.safeApprove(address(metapool), type(uint256).max);
+    }
+
+    /**
+     * @dev Get the index of the coin
+     */
+    function _getMetapoolCoinIndex(address _asset)
+        internal
+        view
+        returns (uint128)
```

Comment:
If this returns an int128, then all the places that use this function will not need to do a cast on the result.

Reply:
we can change this to int128, but then we need casts on all the places that expect it to be uint256. like: 
```
- metapool.balances(crvCoinIndex) ( 3 occurrences)
- IVault(vaultAddress).burnForStrategy(_removedAmounts[mainCoinIndex])
- IVault(vaultAddress).burnForStrategy(_removedAmounts[mainCoinIndex])
```
Looks like uint128 is implicitly cast to uint256 (in all above places) while int128 needs to be explicitly cast. 

I would rather leave it as is.

Reply:
I'm good leaving it as it is.